### PR TITLE
add button to view haskell code of fully defined functions

### PR DESCRIPTION
**Pull request by lstreckeisen**

### CODE REVIEW

File: backend/src/VFP/Translation/TranslateToHaskellSource.hs [hunk 1]
```diff
@@ -54,3 +55,18 @@ translateTypedToHaskell (TypedReference typ name args) =
 translateTypedToHaskell (TypedLiteral typ name) =
     "TypedLiteral " ++ brace (printHaskellType typ) ++ space ++ quote name
 
+translateToHaskellCode :: TypedValue -> String
+translateToHaskellCode (TypedValueDefinition typ name inner) = name ++ " :: " ++ (translateTypeToHaskellType typ) ++ "\n" ++ name ++ " = " ++ (translateToHaskellCode inner)
+translateToHaskellCode (TypedTypeHole _ _) = error "No type holes allowed"
+translateToHaskellCode (TypedLambda _ (_, pname) body) = brace $ "\\" ++ pname ++ " -> " ++ translateToHaskellCode body
+translateToHaskellCode (TypedReference _ "if" [condition, thenBody, elseBody]) = "if " ++ translateToHaskellCode condition ++ "\n\tthen " ++ translateToHaskellCode thenBody ++ "\n\telse " ++ translateToHaskellCode elseBody
+translateToHaskellCode (TypedReference _ name args) = name ++ " " ++ intercalate " " (map translateToHaskellCode args)
```
Comment (fl3pp) [inline backend/src/VFP/Translation/TranslateToHaskellSource.hs:63]: Isn't there a brace required here? What happens if reference arguments are nested?

File: backend/src/VFP/Translation/TranslateToHaskellSource.hs [hunk 2]
```diff
@@ -54,3 +55,18 @@ translateTypedToHaskell (TypedReference typ name args) =
...
+translateToHaskellCode (TypedLiteral typ name) = case typ of
+    (Primitive _) -> name
+    _ -> error $ "Unsuppported literal type " ++ show typ
+
+translateTypeToHaskellType :: Type -> String
+translateTypeToHaskellType (Primitive n) = n
+translateTypeToHaskellType (Generic num) = [chr (ord 'a' - 1 + num)]
+translateTypeToHaskellType (List inner) = "[" ++ translateTypeToHaskellType inner ++ "]"
+translateTypeToHaskellType (Function from to) = translateTypeToHaskellType from ++ " -> " ++ translateTypeToHaskellType to
```
Comment (fl3pp) [inline backend/src/VFP/Translation/TranslateToHaskellSource.hs:72]: This needs to be nested, you can look into the UIModel how to do it.

Example: A function, that takes a function as first argument

File: frontend/app/VFP/Frontend.hs [hunk 3]
```diff
@@ -145,15 +151,37 @@ createSideBarContainer = UI.new # set UI.id_ "visual-fp-sidebar"
...
 
-createErrorMessage :: String -> UI Element
-createErrorMessage errorMessage = do
-  errorElement <- UI.div #. "error-message"
-  errorMessageElement <- UI.p # set UI.text errorMessage
-  errorDismissButton <- UI.button # set UI.text "X"
-  _ <- element errorElement #+ [element errorMessageElement, element errorDismissButton]
-  on UI.click errorDismissButton $ \_ -> delete errorElement
-  return errorElement
+createPopup :: String -> String -> String -> UI (Element, Event ())
```
[the remark this answers is not in the document]
Comment (lstreckeisen) [reply frontend/app/VFP/Frontend.hs:154]: Yes, that would be the way to go here to create a prompt feature, but building the popup is not the problem.
The "problem"/time-consuming part is getting the value the user entered and putting it in the value definition.

[the remark this answers is not in the document]
Comment (fl3pp) [reply frontend/app/VFP/Frontend.hs:154]: I see 👍

**Review by fl3pp — APPROVED** [review]

**Pull request merged**